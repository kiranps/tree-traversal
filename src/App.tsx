import React, { Component, Fragment } from "react";
import * as d3 from "d3";
import { TREE_DATA } from "./constants";
import {
  TreeBox,
  SelectionBox,
  Select,
  Option,
  TraversedList,
  Item,
  Button
} from "./Styled";
import {
  diagonal,
  inOrderTraversal,
  preOrderTraversal,
  postOrderTraversal,
  breadthFirstTraversal
} from "./helper";

class App extends Component {
  root: any;
  traversal: any;
  arr: any;
  constructor(props: any) {
    super(props);
    const treeLayout = d3.tree().size([320, 200]);
    const root = d3.hierarchy(TREE_DATA);
    treeLayout(root);
    this.root = root;
    this.arr = preOrderTraversal(this.root);
  }

  state = { traversalType: "preorder", traversing: false, tree: [] };

  componentDidMount() {
    const nodes = d3
      .select("svg g.nodes")
      .selectAll("circle.node")
      .data(this.root.descendants())
      .enter()
      .append("g");

    // nodes
    nodes
      .append("circle")
      .attr("stroke", "#555")
      .attr("stroke-opacity", 0.6)
      .attr("stroke-width", "3px")
      .attr("fill", "#fff")
      .attr("cx", (d: any) => d.x)
      .attr("cy", (d: any) => d.y)
      .attr("r", 15);

    // text
    nodes
      .append("text")
      .attr("dx", (d: any) => d.x - 7)
      .attr("dy", (d: any) => d.y + 4)
      .attr("font-size", "12px")
      .attr("fill", "#5f5f5f")
      .attr("text-anchor", "start")
      .text((d: any) => d.value);

    // links
    d3.select("svg g.paths")
      .selectAll("path")
      .data(this.root.links())
      .enter()
      .append("path")
      .attr("d", (d: any) => diagonal(d))
      .attr("fill", "none")
      .attr("stroke", "#555")
      .attr("stroke-opacity", 0.4)
      .attr("stroke-width", "1.5px");
  }

  handleChange = (e: React.FormEvent<HTMLSelectElement>) => {
    let traversalType = e.currentTarget.value;
    this.resetNodes();

    switch (traversalType) {
      case "preorder":
        this.arr = preOrderTraversal(this.root);
        break;
      case "postorder":
        this.arr = postOrderTraversal(this.root);
        break;
      case "inorder":
        this.arr = inOrderTraversal(this.root);
        break;
      case "bft":
        this.arr = breadthFirstTraversal(this.root);
        break;
    }
  };

  resetNodes = () => {
    d3.select("svg g.nodes")
      .selectAll("circle")
      .attr("fill", "#fff")
      .attr("stroke", "#555");

    d3.select("svg g.nodes")
      .selectAll("text")
      .attr("fill", "#5f5f5f");
  };

  toggleTraversal = () => {
    this.state.traversing ? this.stopTraversal() : this.startTraversal();
  };

  stopTraversal = () => {
    clearTimeout(this.traversal);
    this.resetNodes();
    this.setState({ traversing: false, tree: [] });
  };

  startTraversal = () => {
    this.setState({ traversing: true, tree: [] });
    this.animateTraversal(this.arr);
  };

  animateTraversal = (arr: any[]) => {
    if (arr.length > 0) {
      this.traversal = setTimeout(() => {
        const x = arr[0];
        this.setState({ tree: [...this.state.tree, x.value] });
        d3.select("svg g.nodes")
          .selectAll("circle")
          .filter((d: any) => x.x === d.x && x.y === d.y)
          .attr("fill", "#2196F3")
          .attr("stroke", "#2196F3");

        d3.select("svg g.nodes")
          .selectAll("text")
          .filter((d: any) => x.x === d.x && x.y === d.y)
          .attr("fill", "#fff");

        this.animateTraversal(arr.slice(1));
      }, 2000);
    }
  };

  render() {
    const { tree, traversing } = this.state;
    return (
      <Fragment>
        <TreeBox>
          <svg width="200" height="260">
            <g transform="translate(-70, 30)">
              <g className="paths" />
              <g className="nodes" />
            </g>
          </svg>
        </TreeBox>
        <SelectionBox>
          <Select onChange={this.handleChange}>
            <Option value="preorder">preorder</Option>
            <Option value="postorder">postorder</Option>
            <Option value="inorder">inorder</Option>
            <Option value="bft">bft</Option>
          </Select>
          <Button onClick={this.toggleTraversal}>
            {traversing ? "stop" : "start"}
          </Button>
          <TraversedList>
            {tree.map((x, i) => (
              <Item key={i}>{x}</Item>
            ))}
          </TraversedList>
        </SelectionBox>
      </Fragment>
    );
  }
}

export default App;
