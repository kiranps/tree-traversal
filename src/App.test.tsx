import React from "react";
import ReactDOM from "react-dom";
import { mount } from "enzyme";
import App from "./App";

jest.useFakeTimers();

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("should match with snapshot", () => {
  const component = mount(<App />);
  expect(component).toMatchSnapshot();
});

it("should render Tree diagram", () => {
  const component = mount(<App />);
  expect(component.find("Tree").length).toEqual(1);
});

it("should animate the tree in inorder", () => {
  const inorder = ["12", "23", "16", "4", "9", "34", "92"];
  const component = mount(<App />);
  component.find("Select").simulate("change", { target: { value: "inorder" } });
  component.find("Button").simulate("click");
  jest.runAllTimers();
  component.update();
  inorder.forEach((x, i) => {
    expect(
      component
        .find("Item")
        .at(i)
        .text()
    ).toEqual(inorder[i]);
  });
});

it("should animate tree in postorder", () => {
  const postorder = ["12", "16", "9", "4", "23", "92", "34"];
  const component = mount(<App />);
  component
    .find("Select")
    .simulate("change", { target: { value: "postorder" } });
  component.find("Button").simulate("click");
  jest.runAllTimers();
  component.update();
  postorder.forEach((x, i) => {
    expect(
      component
        .find("Item")
        .at(i)
        .text()
    ).toEqual(postorder[i]);
  });
});

it("should animate tree in preorder", () => {
  const preorder = ["34", "23", "12", "4", "16", "9", "92"];
  const component = mount(<App />);
  component
    .find("Select")
    .simulate("change", { target: { value: "preorder" } });
  component.find("Button").simulate("click");
  jest.runAllTimers();
  component.update();
  preorder.forEach((x, i) => {
    expect(
      component
        .find("Item")
        .at(i)
        .text()
    ).toEqual(preorder[i]);
  });
});

it("should animate tree in bft", () => {
  const bft = ["34", "23", "92", "12", "4", "16", "9"];
  const component = mount(<App />);
  component.find("Select").simulate("change", { target: { value: "bft" } });
  component.find("Button").simulate("click");
  jest.runAllTimers();
  component.update();
  bft.forEach((x, i) => {
    expect(
      component
        .find("Item")
        .at(i)
        .text()
    ).toEqual(bft[i]);
  });
});
