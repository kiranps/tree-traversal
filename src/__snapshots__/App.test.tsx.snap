// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match with snapshot 1`] = `
<App>
  <TreeBox>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "TreeBox-gzi78k-0",
            "isStatic": true,
            "lastClassName": "sWcqm",
            "rules": Array [
              "position:absolute;left:0;top:0;bottom:0;width:50%;text-align:center;padding-top:14%;",
            ],
          },
          "displayName": "TreeBox",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "TreeBox-gzi78k-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="TreeBox-gzi78k-0 sWcqm"
      >
        <Tree>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Tree-gzi78k-7",
                  "isStatic": true,
                  "lastClassName": "dZHjRf",
                  "rules": Array [
                    "width:200px;height:260px;",
                  ],
                },
                "displayName": "Tree",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Tree-gzi78k-7",
                "target": "svg",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={
              Object {
                "current": <svg
                  class="Tree-gzi78k-7 dZHjRf"
                >
                  <g
                    transform="translate(-70, 30)"
                  >
                    <g>
                      <path
                        d="M182.85714285714286,0C160,0 160,66.66666666666667 137.14285714285714,66.66666666666667"
                        fill="none"
                        stroke="#555"
                        stroke-opacity="0.4"
                        stroke-width="1.5px"
                      />
                      <path
                        d="M182.85714285714286,0C205.71428571428572,0 205.71428571428572,66.66666666666667 228.57142857142858,66.66666666666667"
                        fill="none"
                        stroke="#555"
                        stroke-opacity="0.4"
                        stroke-width="1.5px"
                      />
                      <path
                        d="M137.14285714285714,66.66666666666667C114.28571428571428,66.66666666666667 114.28571428571428,133.33333333333334 91.42857142857143,133.33333333333334"
                        fill="none"
                        stroke="#555"
                        stroke-opacity="0.4"
                        stroke-width="1.5px"
                      />
                      <path
                        d="M137.14285714285714,66.66666666666667C160,66.66666666666667 160,133.33333333333334 182.85714285714286,133.33333333333334"
                        fill="none"
                        stroke="#555"
                        stroke-opacity="0.4"
                        stroke-width="1.5px"
                      />
                      <path
                        d="M182.85714285714286,133.33333333333334C160,133.33333333333334 160,200 137.14285714285714,200"
                        fill="none"
                        stroke="#555"
                        stroke-opacity="0.4"
                        stroke-width="1.5px"
                      />
                      <path
                        d="M182.85714285714286,133.33333333333334C205.71428571428572,133.33333333333334 205.71428571428572,200 228.57142857142858,200"
                        fill="none"
                        stroke="#555"
                        stroke-opacity="0.4"
                        stroke-width="1.5px"
                      />
                    </g>
                    <g>
                      <g>
                        <circle
                          cx="182.85714285714286"
                          cy="0"
                          fill="#fff"
                          r="15"
                          stroke="#555"
                          stroke-opacity="0.6"
                          stroke-width="3px"
                        />
                        <text
                          dx="175.85714285714286"
                          dy="4"
                          fill="#5f5f5f"
                          font-size="12px"
                        >
                          34
                        </text>
                      </g>
                      <g>
                        <circle
                          cx="137.14285714285714"
                          cy="66.66666666666667"
                          fill="#fff"
                          r="15"
                          stroke="#555"
                          stroke-opacity="0.6"
                          stroke-width="3px"
                        />
                        <text
                          dx="130.14285714285714"
                          dy="70.66666666666667"
                          fill="#5f5f5f"
                          font-size="12px"
                        >
                          23
                        </text>
                      </g>
                      <g>
                        <circle
                          cx="228.57142857142858"
                          cy="66.66666666666667"
                          fill="#fff"
                          r="15"
                          stroke="#555"
                          stroke-opacity="0.6"
                          stroke-width="3px"
                        />
                        <text
                          dx="221.57142857142858"
                          dy="70.66666666666667"
                          fill="#5f5f5f"
                          font-size="12px"
                        >
                          92
                        </text>
                      </g>
                      <g>
                        <circle
                          cx="91.42857142857143"
                          cy="133.33333333333334"
                          fill="#fff"
                          r="15"
                          stroke="#555"
                          stroke-opacity="0.6"
                          stroke-width="3px"
                        />
                        <text
                          dx="84.42857142857143"
                          dy="137.33333333333334"
                          fill="#5f5f5f"
                          font-size="12px"
                        >
                          12
                        </text>
                      </g>
                      <g>
                        <circle
                          cx="182.85714285714286"
                          cy="133.33333333333334"
                          fill="#fff"
                          r="15"
                          stroke="#555"
                          stroke-opacity="0.6"
                          stroke-width="3px"
                        />
                        <text
                          dx="175.85714285714286"
                          dy="137.33333333333334"
                          fill="#5f5f5f"
                          font-size="12px"
                        >
                          4
                        </text>
                      </g>
                      <g>
                        <circle
                          cx="137.14285714285714"
                          cy="200"
                          fill="#fff"
                          r="15"
                          stroke="#555"
                          stroke-opacity="0.6"
                          stroke-width="3px"
                        />
                        <text
                          dx="130.14285714285714"
                          dy="204"
                          fill="#5f5f5f"
                          font-size="12px"
                        >
                          16
                        </text>
                      </g>
                      <g>
                        <circle
                          cx="228.57142857142858"
                          cy="200"
                          fill="#fff"
                          r="15"
                          stroke="#555"
                          stroke-opacity="0.6"
                          stroke-width="3px"
                        />
                        <text
                          dx="221.57142857142858"
                          dy="204"
                          fill="#5f5f5f"
                          font-size="12px"
                        >
                          9
                        </text>
                      </g>
                    </g>
                  </g>
                </svg>,
              }
            }
          >
            <svg
              className="Tree-gzi78k-7 dZHjRf"
            />
          </StyledComponent>
        </Tree>
      </div>
    </StyledComponent>
  </TreeBox>
  <SelectionBox>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "SelectionBox-gzi78k-1",
            "isStatic": true,
            "lastClassName": "drXKbc",
            "rules": Array [
              "position:absolute;right:0;top:0;bottom:0;width:50%;background-color:#eee;text-align:center;padding-top:20%;",
            ],
          },
          "displayName": "SelectionBox",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "SelectionBox-gzi78k-1",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="SelectionBox-gzi78k-1 drXKbc"
      >
        <Select
          onChange={[Function]}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Select-gzi78k-2",
                  "isStatic": true,
                  "lastClassName": "jYZjfo",
                  "rules": Array [
                    "width:200px;height:40px;text-align:center;background-color:#fff;outline:none;margin:10px auto;padding:5px 10px;border:1px solid #d6d6d6;color:#5f5f5f;",
                  ],
                },
                "displayName": "Select",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Select-gzi78k-2",
                "target": "select",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onChange={[Function]}
          >
            <select
              className="Select-gzi78k-2 jYZjfo"
              onChange={[Function]}
            >
              <Option
                value="preorder"
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "Option-gzi78k-3",
                        "isStatic": true,
                        "lastClassName": "kxuIac",
                        "rules": Array [
                          "color:#5f5f5f;",
                        ],
                      },
                      "displayName": "Option",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "Option-gzi78k-3",
                      "target": "option",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  value="preorder"
                >
                  <option
                    className="Option-gzi78k-3 kxuIac"
                    value="preorder"
                  >
                    preorder
                  </option>
                </StyledComponent>
              </Option>
              <Option
                value="postorder"
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "Option-gzi78k-3",
                        "isStatic": true,
                        "lastClassName": "kxuIac",
                        "rules": Array [
                          "color:#5f5f5f;",
                        ],
                      },
                      "displayName": "Option",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "Option-gzi78k-3",
                      "target": "option",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  value="postorder"
                >
                  <option
                    className="Option-gzi78k-3 kxuIac"
                    value="postorder"
                  >
                    postorder
                  </option>
                </StyledComponent>
              </Option>
              <Option
                value="inorder"
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "Option-gzi78k-3",
                        "isStatic": true,
                        "lastClassName": "kxuIac",
                        "rules": Array [
                          "color:#5f5f5f;",
                        ],
                      },
                      "displayName": "Option",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "Option-gzi78k-3",
                      "target": "option",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  value="inorder"
                >
                  <option
                    className="Option-gzi78k-3 kxuIac"
                    value="inorder"
                  >
                    inorder
                  </option>
                </StyledComponent>
              </Option>
              <Option
                value="bft"
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "Option-gzi78k-3",
                        "isStatic": true,
                        "lastClassName": "kxuIac",
                        "rules": Array [
                          "color:#5f5f5f;",
                        ],
                      },
                      "displayName": "Option",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "Option-gzi78k-3",
                      "target": "option",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  value="bft"
                >
                  <option
                    className="Option-gzi78k-3 kxuIac"
                    value="bft"
                  >
                    bft
                  </option>
                </StyledComponent>
              </Option>
            </select>
          </StyledComponent>
        </Select>
        <Button
          onClick={[Function]}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Button-gzi78k-6",
                  "isStatic": true,
                  "lastClassName": "gwyvcj",
                  "rules": Array [
                    "margin-left:10px;border:1px solid #d6d6d6;color:#5f5f5f;background-color:#fff;border-radius:3px;height:40px;cursor:pointer;outline:none;min-width:100px;user-select:none;",
                  ],
                },
                "displayName": "Button",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Button-gzi78k-6",
                "target": "button",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onClick={[Function]}
          >
            <button
              className="Button-gzi78k-6 gwyvcj"
              onClick={[Function]}
            >
              start
            </button>
          </StyledComponent>
        </Button>
        <TraversedList>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "TraversedList-gzi78k-4",
                  "isStatic": true,
                  "lastClassName": "eUAOUJ",
                  "rules": Array [
                    "margin-top:10px;",
                  ],
                },
                "displayName": "TraversedList",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "TraversedList-gzi78k-4",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="TraversedList-gzi78k-4 eUAOUJ"
            />
          </StyledComponent>
        </TraversedList>
      </div>
    </StyledComponent>
  </SelectionBox>
</App>
`;
